### PR TITLE
Added storage image support.

### PR DESCRIPTION
**Pull request by asuonpaa**
Fixes https://github.com/google/amber/issues/327

### CODE REVIEW

File: src/vulkan/buffer_descriptor.h [hunk 1]
```diff
@@ -31,14 +32,9 @@ namespace vulkan {
...
 
-enum class DescriptorType : uint8_t {
-  kStorageBuffer = 0,
-  kUniformBuffer,
-};
-
 /// Stores descriptor set and binding information for storage and uniform
 /// buffers.
-class BufferDescriptor {
+class BufferDescriptor : public Descriptor {
```
Comment (dj2) [inline src/vulkan/buffer_descriptor.h]: How difficult would it be to put the changes for splitting out the Descriptor class into a separate CL which is mostly just moving code around? (Similar for, I think, the image resource code that was split out? We should be able to split and land it pretty quickly.

Comment (asuonpaa) [reply src/vulkan/buffer_descriptor.h]: Most likely it sounds easier than it is and would require some kind of intermediate code that would then be discarded for the final result. I'd rather get the storage image support into good shape quickly and then add other descriptor types one by one as separate PRs.

File: src/vulkan/pipeline.cc [hunk 2]
```diff
@@ -296,15 +304,17 @@ Result Pipeline::AddDescriptor(const BufferCommand* cmd) {
...
-  auto* buf_desc = static_cast<BufferDescriptor*>(desc);
-  if (cmd->GetValues().empty()) {
-    Result r = buf_desc->SetSizeInElements(cmd->GetBuffer()->ElementCount());
-    if (!r.IsSuccess())
-      return r;
-  } else {
-    Result r = buf_desc->AddToBuffer(cmd->GetValues(), cmd->GetOffset());
-    if (!r.IsSuccess())
-      return r;
+  if (desc->IsStorageBuffer() || desc->IsUniformBuffer()) {
```
Comment (dj2) [inline src/vulkan/pipeline.cc]: Is the assumption here that htere is no buffer data for image buffers? I don't think that's a valid assumption to make as we may have image data input, yea?

File: src/amberscript/parser.cc [hunk 3]
```diff
@@ -705,6 +707,8 @@ Result Parser::ParsePipelineBind(Pipeline* pipeline) {
       Result r = pipeline->SetPushConstantBuffer(buffer);
       if (!r.IsSuccess())
         return r;
+    } else if (token->AsString() == "storage_image") {
```
Comment (dj2) [inline src/amberscript/parser.cc]: Can you add a parser unit test for storage_image?

File: src/amberscript/parser_bind_test.cc [hunk 4]
```diff
@@ -1325,5 +1325,121 @@ END)";
...
+BUFFER texture FORMAT R8G8B8A8_UNORM
+
+PIPELINE compute pipeline
+  ATTACH compute_shader
+  BIND BUFFER texture AS storage_image DESCRIPTOR_SET 0 BINDING 0
+END)";
+
+  Parser parser;
+  Result r = parser.Parse(in);
+  ASSERT_TRUE(r.IsSuccess());
```
Comment (dj2) [inline src/amberscript/parser_bind_test.cc]: Can you also pull out the buffer and verify it's set as `storage_image`?

Comment (dj2) [inline src/amberscript/parser_bind_test.cc]: ASSERT_TRUE(r.IsSuccess()) << r.Error();

File: src/amberscript/parser_bind_test.cc [hunk 5]
```diff
@@ -1325,5 +1325,121 @@ END)";
...
+PIPELINE graphics pipeline
+  ATTACH vert_shader
+  ATTACH frag_shader
+  BIND BUFFER texture AS storage_image DESCRIPTOR_SET 0 BINDING 0
+  BIND BUFFER framebuffer AS color LOCATION 0
+END)";
+
+  Parser parser;
+  Result r = parser.Parse(in);
+  ASSERT_TRUE(r.IsSuccess());
```
Comment (dj2) [inline src/amberscript/parser_bind_test.cc]: ditto

File: src/vulkan/descriptor.cc [hunk 6]
```diff
@@ -44,11 +44,9 @@ VkDescriptorType Descriptor::GetVkDescriptorType() const {
       return VK_DESCRIPTOR_TYPE_STORAGE_IMAGE;
     case DescriptorType::kCombinedImageSampler:
       return VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER;
-    case DescriptorType::kSampledImage:
-      return VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE;
     default:
-      assert(false && "unexpected descriptor type");
-      return VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER;
+      assert(type_ == DescriptorType::kSampledImage);
```
Comment (dj2) [inline src/vulkan/descriptor.cc]: Why make this change instead of what was here previously? I think the original was clearer.

Comment (asuonpaa) [reply src/vulkan/descriptor.cc]: CI clang build was failing because of having default even when all enums where covered.

**Comment by dj2** [comment]
Looks like you need to add your new files to the Makefile and the BUILD.gn files.

**Comment by asuonpaa** [comment]
> Looks like you need to add your new files to the Makefile.

Which makefile? Those are already in CMakeLists.txt.

**Pull request merged**